higher/index.ts

```diff
@@ -0,0 +1,2314 @@
+      groupID: groupID,
+      data: data,
+    }, this.#resolveIDs([groupID]));
+  }
+  
+  /**
+   * Deletes data key (and associated value).
+   *
+   * @param {string} key data key
+   *
+   * @private
+   */
+  #deleteDataLocally({ key }: { key: string }) {
+    this.localStorageWrapper.remove(key);
+  }
+  
+  async #deleteDataRemotely(key: string, idkeys: string[]) {
+    await this.core.sendMessage(idkeys, {
+      msgType: Higher.DELETE_DATA,
+      key: key,
+    });
+  }
+  
+  async #deleteData(key: string, idkeys: string[]) {
+    await this.#deleteDataRemotely(key, idkeys);
+    // TODO impl pending state
+  }
+  
+  /**
+   * Function that handles propagating data removal to all members of the 
+   * group and deleting the datum locally.
+   *
+   * @param {string} key data key
+   * @param {?string} groupID group ID to use for resolving idkeys to delete from
+   *
+   * @private
+   */
+  async #removeDataHelper(
+      key: string,
+      curGroupID: string = null,
+      toUnshareGroupID: string = null
+  ) {
+    if (curGroupID === null) {
+      curGroupID = this.localStorageWrapper.get(key)?.groupID;
+    }
+    if (curGroupID !== null) {
+      let idkey = this.core.olmWrapper.getIdkey();
+      if (!this.#hasWriterPriv(idkey, curGroupID)) {
+        this.#printBadDataPermissionsError();
+        return;
+      }
+  
+      // delete data from select devices only (unsharing)
+      if (toUnshareGroupID !== null) {
+        await this.#deleteData(key, this.#resolveIDs([toUnshareGroupID]));
+        return;
+      }
+  
+      // delete data from all devices in group including current (removing data)
+      await this.#deleteData(key, this.#resolveIDs([curGroupID]));
+    }
+  }
+
+  /**
+   * Validates a message via the validateCallback, which can either be:
+   * defined by the application (through setValidateCallback()), or
+   * the default callback (defaultValidateCallback).
+   *
+   * @param {string} payload the decrypted message contents
+   *
+   * @private
+   */
+  #validate(payload: payloadType): boolean {
+    return this.validateCallback(payload);
+  }
+
+  /**
+   * Sets the callback function with which to perform message validation
+   * for this application.
+   *
+   * TODO validate newValidateCallback to ensure it only takes one arg (payload)
+   * and returns boolean.
+   *
+   * @param {callback} newValidateCallback new validation callback
+   */
+  setValidateCallback(callback: (payloadType) => boolean) {
+    this.validateCallback = callback;
+  }
+
+  /**
+   * Generates ID for data value, resolves the full key given the prefix, 
+   * adds group information, stores value and sends to other devices
+   * in the group to also store.
+   * TODO allow named groups from app.
+   *
+   * @param {string} prefix prefix name
+   * @param {Object} data app-specific data object
+   * @param {string} id app-specific object id
+   */
+  async setData(prefix: string, id: string, data: any) {
```

Comment:
could just be key value params, fixing it here would technically resolve above comment